contributors/design-proposals/loadbalancer-proposal.md

```diff
@@ -0,0 +1,396 @@
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Ingress, LoadBalancer and LoadBalancerClaim proposal
+
+**Author**: @mqliang, @ddysher
+
+Nov 2016
+
+## Overview
+
+The proposal aims to define a set of APIs to abstract loadbalancer resource,
+similar to how PV/PVC works. In the end, we'll
+
+* introduce a new API object `LoadBalancer` to represent a loadbalancer (think of PV)
+* introduce a new API object `LoadBalancerClaim` to claim a new or existing
+  LoadBalancer (think of PVC)
+* change existing Ingress API to include a `LBSource` field, which either directly
+  uses a LoadBalancer, or via a LoadBalancerClaim (think of Pod using PV or PVC)
+
+We will also introduce network resource such as bandwidth and iops. LoadBalancerClaim
+uses these attributes to claim a LoadBalacer. Though similar to PV/PVC, it is
+important to know that LoadBalancer and LoadBalancerClaim is not 1-to-1 binding;
+rather, it is a 1-to-many relationship. That is, a LoadBalancer can be used to
+serve multiple LoadBalancerClaim.
+
+## Background
+
+### Current Ingress behavior
+
+Ingress can be used to expose a service in the kubernetes cluster:
+
+* cluster admin deploys an ingress-controller Pod beforehand
+* user creates Ingress resource
+* the ingress-controller Pod list&watch **All** Ingress Resources in the cluster,
+  when it sees a new Ingress resource:
+  * on cloud provider, it calls the cloud provider to sync the ingress L7
+    loadbalancing rules
+  * on bare-metal, it syncs nginx (or haproxy, etc) config then reload
+* user out of cluster can then access service in the cluster:
+  * on bare-metal, accessing the node's ip on which ingress-controller Pod is
+    running, ingress-controller Pod will forward request into cluster based on
+    rules defined in Ingress resource
+  * on cloud-provider, accessing the ip provided by cloud provider loadbalancer,
+    cloud provider will forward request into cluster based on rules defined in
+    Ingress Resource
+
+### Limitations of current Ingress implementation
+
+* How many ingress-controller Pods should run in a cluster? Should all
+  ingress-controller Pod list&watch all Ingress resource? There is no way to
+  bind or schedule Ingress resource to a ingress-controller Pod, which result in:
```

Comment:
One Ingress controller handles the setup for all services of a certain class in the cluster. I understand why you would run multiple kinds of Ingress controllers (one for handling each type), but why would we run more than that?

Reply:
It mainly because a single L7 loadbalancer may have a upper limit about some resource(e.g bandwidth). For example, a single L7 loadbalancer on Aliyun have a upper limit of 0.5G bandwidth. It's impossible let a single ingress controller to serve all ingress.

So, instead of let a single "Aliyun" class ingress-controller to serve all "Aliyun" class ingress. I want deploy multiple "Aliyun" class ingress-controller, and have some "scheduling" or "binding" logic to assign ingress to ingress-controller.

Reply:
It seems like an implementation detail as to how many pods are used to implement the controller logic, it may not be 1-1 as you are suggesting here. It's not clear to me why we should expose this to the user of k8s.

We should also be careful to distinguish between "ingress" spec which describes a L7 exposed service and the actual LBs that implement the service. 

Does each Aliyun LB Ingress have a different IP? Or do they share the same IP to the external user?

Reply:
@bowei 
> Does each Aliyun LB Ingress have a different IP? Or do they share the same IP to the external user?

Yes, each Aliyun L7 LB have a different IP. See documents here https://www.alibabacloud.com/help/doc-detail/32459.htm?spm=a3c0i.o32460en.b99.6.387f4cf8yNv1ao

I understand that aws/gce LB is very powerful, basically you can send as many traffic to LB as possible and LB itself will scale up/down accordingly. There is no maximum bandwidth defined for a single LB on aws/gce. 

However, on other cloud platform and bare-metal , L7 LB itself is not horizontally scalable